float64 support in FIL functions

canonizer opened the pull request:
Templatized functions related to FIL inference in preparation of `float64` support.

Instantiations of templates with `float64`, or tests for `float64`, _are not included_; they will be provided in a future pull request.

This is pull request 2 of 3 to integrate https://github.com/rapidsai/cuml/pull/4646. This pull request is partly based on the work by @levsnv.

wphicks commented on cpp/src/fil/common.cuh:
```diff
@@ -46,8 +46,9 @@ struct storage_base {
-struct tree<dense_node<real_t>> : tree_base {
+template <typename real_t_>
+struct tree<dense_node<real_t_>> : tree_base {
+  using real_t = real_t_;
```
I'd slightly prefer we spell out member types as e.g. `real_type` for better consistency with STL conventions, but let's not hold up the PR if that's going to be a difficult refactor.

canonizer replied:
Done (in type-scope using declarations).

wphicks commented on cpp/src/fil/fil.cu:
```diff
@@ -35,35 +35,37 @@ creation and prediction (the main inference kernel is defined in infer.cu). */
 namespace ML {
 namespace fil {
 
+__host__ __device__ double sigmoid(double x) { return 1.0 / (1.0 + exp(-x)); }
```
Any reason not to templatize these two functions?

canonizer replied:
Done (templatized).

wphicks commented on cpp/src/fil/fil.cu:
```diff
@@ -129,7 +134,32 @@ struct forest {
+  int depth_       = 0;
+  algo_t algo_     = algo_t::NAIVE;
+  output_t output_ = output_t::RAW;
+  shmem_size_params class_ssp_, proba_ssp_;
```
Let's split this into two declarations.

canonizer replied:
Done.

wphicks commented on cpp/src/fil/fil.cu:
```diff
@@ -139,8 +169,8 @@ struct forest {
     output_                          = params->output;
-    threshold_                       = params->threshold;
-    global_bias_                     = params->global_bias;
+    threshold_                       = static_cast<real_t>(params->threshold);
```
Not necessarily saying we should, but would it be possible to templatize `params` here such that the threshold is already of the correct type?

canonizer replied:
I made those parameters `double` in `forest_params` and corresponding test parameters. This looks simpler than templatizing them, and they are not used in any performance-critical code.

wphicks commented on cpp/src/fil/fil.cu:
```diff
@@ -173,8 +203,21 @@ struct forest {
 
-  void predict(
-    const raft::handle_t& h, float* preds, const float* data, size_t num_rows, bool predict_proba)
+  virtual void predict(const raft::handle_t& h,
```
I'd strongly recommend we reconsider this structure. There is no reason that the downstream caller needs to have a single `predict` API with type erasure for both `float` and `double`. I also want to make sure we are absolutely confident on what the performance implications will be of making this a virtual function, particularly for applications that are making many repeated calls to `predict` in a short period.

canonizer replied:
Done.

wphicks commented on cpp/src/fil/fil.cu:
```diff
@@ -305,7 +350,7 @@ struct forest {
         num_values_to_transform,
         ot,
-        num_trees_ > 0 ? (1.0f / num_trees_) : 1.0f,
+        num_trees_ > 0 ? (real_t(1) / num_trees_) : real_t(1),
```
In general, we should probably [prefer](https://herbsutter.com/2013/05/09/gotw-1-solution/) uniform initialization syntax (e.g. `real_t{1}`), but no need to go crazy about fixing every instance of it.

canonizer replied:
I have a strong preference for `()` in constructor calls and constant conversions. I prefer to reserve `{}` to struct and array literals, as well as initializer lists.

wphicks commented on cpp/src/fil/fil.cu:
```diff
@@ -374,14 +407,14 @@ struct dense_forest<dense_node<real_t>> : forest {
     */
     int global_node = 0;
-    for (int tree = 0; tree < num_trees_; ++tree) {
+    for (int tree = 0; tree < this->num_trees_; ++tree) {
```
Why is `this->` necessary here?

canonizer replied:
Short answer: without `this->`, it doesn't compile (`num_trees` becomes an undefined identifier). This also applies to other instances of `this->` in this file.

Long answer: I think this has something to do with template-dependent lookup. I presume that as the base type is parameterized with `real_t`, which is a type parameter in this type, all members there become type-dependent, and are not found with a simple name lookup.

I think https://en.cppreference.com/w/cpp/language/dependent_name explains why, but I admit that I myself don't completely understand this. If you know a better way to fix this, please let me know.

wphicks commented on cpp/src/fil/fil.cu:
```diff
@@ -583,59 +619,56 @@ void check_params(const forest_params_t* params, bool dense)
   delete f;
 }
 
+/// part of C API - preds and data are either both pointers to float or to double
```
I would recommend having a separate function for the C API that includes the necessary type dispatch and maintain type safety for C++ callers.

canonizer replied:
Done (removed the comment).

wphicks commented on cpp/src/fil/infer.cu:
```diff
@@ -372,18 +369,22 @@ __device__ __forceinline__ void write_best_class(
 /// needed for softmax
-__device__ float shifted_exp(float margin, float max) { return expf(margin - max); }
+struct shifted_exp {
+  __device__ double operator()(double margin, double max) const { return exp(margin - max); }
```
Any reason not to templatize here?

canonizer replied:
Done (templatized).

wphicks commented on cpp/src/fil/infer.cu:
```diff
@@ -792,8 +794,9 @@ template <int NITEMS,
+  using real_t = typename storage_type::real_t;
   extern __shared__ char smem[];
-  float* sdata       = (float*)smem;
+  real_t* sdata      = (real_t*)smem;
```
Let's avoid C-style casts if we're already updating this line.

canonizer replied:
Done.

wphicks commented on cpp/src/fil/infer.cu:
```diff
@@ -845,21 +849,34 @@ __global__ void infer_k(storage_type forest, predict_params params)
+      return ssp
+        .template get_smem_footprint<KernelParams::N_ITEMS, double, KernelParams::LEAF_ALGO>();
+    default:
+      ASSERT(ssp.sizeof_real == 4 || ssp.sizeof_real == 8,
```
Rather than asserting a condition already checked by the switch, let's just do the necessary error handling. If we later add additional cases, we'll have to update 3 places with the current structure.

canonizer replied:
Done.

wphicks commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -674,6 +680,8 @@ void from_treelite(const raft::handle_t& handle,
       "This may lead to predictions with reduced accuracy.");
   }
+  // same as std::common_type: float+double=double, float+int64_t=float
+  typedef decltype(threshold_t{} + leaf_t{}) real_t;
```
Let's use `using` where possible to avoid having to think about if a type alias will work for all template applications.

canonizer replied:
Done.

divyegala commented on cpp/src/fil/internal.cuh:
```diff
@@ -242,7 +242,7 @@ struct node_traits {
   static const bool IS_DENSE = false;
   static const storage_type_t storage_type_enum =
-    std::is_same<sparse_node16<typename node_t::real_t>, node_t>() ? SPARSE : SPARSE8;
+    std::is_same<sparse_node16<real_t>, node_t>() ? SPARSE : SPARSE8;
```
This seems compile-time to me? In that case I'd recommend using `constexpr` which guarantees substitution at compile-time

canonizer replied:
Done (`static constexpr storage_type_enum`).

divyegala commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -655,17 +655,23 @@ void convert(const raft::handle_t& handle,
+constexpr bool type_supported()
+{
+  // not using std::is_floating_point because we did not instantiate fp16-based nodes/trees/forests
+  return std::is_same<real_t, float>() || std::is_same<real_t, double>();
```
This is not a required change, but personally I find `std::is_same_v` more readable than an implicit conversion to `operator bool`

canonizer replied:
Done.

divyegala reviewed (commented):
Some comments/suggestions. Overall looks good to me

canonizer commented on cpp/src/fil/infer.cu:
```diff
@@ -822,7 +826,7 @@ __global__ void infer_k(storage_type forest, predict_params params)
          Same with thread_num_rows > 0
       */
-      typedef typename leaf_output_t<leaf_algo, float>::T pred_t;
+      typedef typename leaf_output_t<leaf_algo, real_t>::T pred_t;
```
Done.

canonizer commented on cpp/include/cuml/fil/fil.h:
```diff
@@ -128,17 +128,19 @@ void free(const raft::handle_t& h, forest_t f);
              forest_t f,
-             float* preds,
-             const float* data,
+             void* preds,
```
Done.

canonizer commented:
Please ignore `python/cuml/decomposition/{pca_mg.pyx,tsvd_mg.pyx,utils.pxd,utils_mg.pxd}`; I've added them from https://github.com/rapidsai/cuml/pull/4645 to get Python wrapper compile.

canonizer commented:
@wphicks @divyegala I've addressed all of your comments, and also templatized methods and types in `fil.h` as we discussed. Could you take another look?

Please ignore the files in `python/cuml/decomposition`; they come from https://github.com/rapidsai/cuml/pull/4645, and are only necessary to fix Cython build errors.

divyegala reviewed (approved):
LGTM

wphicks commented on cpp/src/fil/infer.cu:
```diff
@@ -36,7 +36,7 @@
 #endif  // CUDA_PRAGMA_UNROLL
 
-#define INLINE_CONFIG __forceinline__
+#define INLINE_CONFIG __noinline__
```
Was this intended to be reverted before merger?

canonizer replied:
Fixed.

Thanks for catching it!

canonizer commented:
rerun tests

canonizer commented:
rerun tests

dantegd commented:
rerun tests

dantegd commented:
@gpucibot merge
